KAS-3658: Source file as main file

sergiofenoll opened the pull request:
- Rename usage of `primarySource` to `source` and `primarySourceOf` to `derived`
- Consider `source-file` as main file, so use checks for `derived` instead of `source`
- Fix bug with replacing files. We should first replace the file, then delete. Because if you delete it first and then try to replace a link to the file, `mu-authorization` will throw errors (because it can't find type info about the resource you're overwriting, I think)
- ~~Don't show file extension unless it's explicit which file you're talking about. In `document-card` it's somewhat vague, so leave out the extension there.~~ Always show the extension of the "main file" (derived file if it exists, otherwise source file)

Jenkins run(s):
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/323/~~
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/324/~~
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/325/~~
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/61/~~
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/62/~~
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/65/~~
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/66/~~
- ~~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/67/~~ Flaky test? Can't immediately repro the failing test locally
- http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/68/ :heavy_check_mark: all green :)

Related PRs:
- [x] https://github.com/kanselarij-vlaanderen/app-kaleidos/pull/335
- [x] https://github.com/kanselarij-vlaanderen/yggdrasil/pull/49
- [x] https://github.com/kanselarij-vlaanderen/file-bundling-job-creation-service/pull/3
- [x] https://github.com/kanselarij-vlaanderen/themis-export-service/pull/11

sergiofenoll committed:
feat: rename primarySource(of) to source and derived

sergiofenoll committed:
chore: hide extension of source and derived file unless explicit

Showing .docx when the derived is a .pdf and we open the PDF gets a
littel confusing. We just always show the piece name, and when it's
explicit what file we're working with (e.g. in the details panel, where
you replace the source file only), we show the appropriate extension.

sergiofenoll commented on app/components/documents/document-card.hbs:
```diff
@@ -23,34 +23,34 @@
-                  </AuLinkExternal>
+                  {{#if this.piece.file.derived}}
+                    {{t "as"}}
+                    {{!-- TODO: AuLinkExternal has a font-size of 16px (because it uses the same size as buttons), but that's larger than normal text (which uses 14px).
```
@brenner-company

Dit ziet er zo uit
![image](https://user-images.githubusercontent.com/22411874/200867779-0e079dc1-33cf-4881-aad8-2be11376a46b.png)

`bronbestand` is nogal groter dan de rest van de tekst. Dit komt volgens mij door [deze CSS rules](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/styles/au/_au-components.scss#L202-L206). Misschien is het interessant om de font-size enkel aan te passen voor `au-c-link.button-` klasses. Dus een gewone `AuLinkExternal` heeft wel dezelfde grootte als normale tekst, maar `AuLinkExternal` met een button skin is dan weer groter.

brenner-company replied:
@sergiofenoll Goed opgemerkt 👍 Ik ga dit de volgende sprint opnemen want ik ga goed alles moeten overlopen om te weten dat er geen externe impact is.

sergiofenoll committed:
ci: use Yggdrasil feature image

sergiofenoll committed:
chore: show extension again

If there's a derived file, show the derived extension, otherwise show
the source extension. I suspect this will break the tests because at
first render we show the source extension (because the derived file is
still in-flight).
But if it's a non-issue I will leave it as is, because checking for
the loading state of a relation is kind of annoying in Ember.

sergiofenoll commented on app/components/documents/document-card.hbs:
```diff
@@ -23,34 +23,36 @@
+              class="auk-h4"
+              data-test-document-card-name-value
+            >
+              {{#let (or this.piece.file.derived.extension this.piece.file.extension) as |extension|}}
```
This results in an annoying re-rendering. Because `this.piece.file` loads in before `this.piece.file.derived`, we will always show the source extension before the derived extension, even if derived exists. So it's going to be `document name.docx` then `document name.pdf` for basically all files from now on.

Perhaps a task or resource could help here, just so that the data for `this.piece.file` and `this.piece.file.derived` exist before we start rendering. But I think that would mean breaking up the component some more (the previous versions would need to live in a separate component to make the loading tasks work as expected). Not insurmountable, but a big chunk of work just to show an extension without annoying redrawing.*

*: I could be wrong, I haven't tried yet, maybe it's actually quite easy to work around the redrawing.

erikap replied:
The [`is-pending` helper of ember-promise-helpers](https://github.com/fivetanley/ember-promise-helpers) might be useful to postpone rendering in the template until the relation is loaded, but I'm not agains extracting the history entries to a separate component.

erikap replied:
@sergiofenoll I forgot about it, but I've created a separate component already in [feature/KAS-3505-draft-access-level-pills-bis ](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1526). Triggered Jenkins once more for that branch, but hope to merge it today.

erikap replied:
https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1526 is merged. Maybe best to merge development in this PR before making changes.

sergiofenoll committed:
fix: also delete the derived file when deleting a piece

sergiofenoll committed:
test: update new-document-widget spec for file conversion compatibility

erikap commented on app/services/file-service.js:
```diff
@@ -59,6 +59,8 @@ export default class FileService extends Service {
       return;
     }
+    const derivedFileToDelete = await file.derived;
+    await derivedFileToDelete?.destroyRecord();
```
Nice catch. But `destroyRecord()` on a file is forwarded to the file-service by the dispatcher. This service only removes the triples it has inserted on upload, hence there will be dangling triples `<derived-file> prov:hadPrimarySource <source-file>` in the triplestore if we remove both files like this.
We can work around it by first fetching the derived file and setting/saving `fileToDelete.derived = null`  if it exists. Afterwards, both files (source and derived) can be removed in parallel.

erikap committed:
Remove unused convertedFile relation from piece model

erikap commented on app/components/documents/document-details-panel.js:
```diff
@@ -48,11 +48,10 @@ export default class DocumentsDocumentDetailsPanel extends Component {
-      yield oldSourceFile?.destroyRecord();
-      file.primarySource = this.replacementSourceFile;
-      yield file.save();
+      const oldFile = yield this.args.piece.file;
+      yield oldFile.destroyRecord();
+      this.args.piece.file = this.replacementSourceFile;
+      yield this.args.piece.save();
```
We need to relink the existing derived file (if any) to the replacement source file here.

sergiofenoll replied:
Ah, I was working on the assumption that if you replace the source file, the derived file is "invalid" (but I should also delete that in that case, which I forgot).

Is the derived file still valid if the source file changes?

erikap replied:
Not sure. Let's keep it as you implemented it now (not destroying the derived file) and see what feedback we get from the users.

erikap commented on app/components/documents/document-card/edit-modal.hbs:
```diff
@@ -10,7 +10,11 @@
+        @badgeIcon={{
+          if (eq @piece.file.extension "pdf") "pdf"
+          (if (eq @piece.file.extension "docx") "word"
+          "document")
```
Can we create a helper for this?

erikap commented on app/components/documents/document-card/edit-modal.js:
```diff
@@ -69,22 +69,24 @@ export default class DocumentsDocumentCardEditModalComponent extends Component {
       yield oldFile.destroyRecord();
       this.args.piece.file = this.replacementFile;
```
We need to relink the existing derived file (if any) to the replacement source file here

erikap commented on app/components/documents/document-card/edit-modal.js:
```diff
@@ -11,13 +11,13 @@ export default class DocumentsDocumentCardEditModalComponent extends Component {
   @tracked name;
-  @tracked uploadedSourceFile;
+  @tracked uploadedDerivedFile;
   @tracked replacementFile;
```
Can we consistently use `source` and `derived` in all properties/methods of this component to avoid confusion?

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3658-source-file-as-main-file

sergiofenoll committed:
chore: ensure files are loaded before showing names in document card

This prevents quick flashes of content when suddently the derived file
loads in after we've already displayed the source file's extension.

sergiofenoll committed:
fix: null derived file before deleting to prevent dangling triples

sergiofenoll committed:
feat: add document-icon helper

It returns the icon name based on the file extension

sergiofenoll committed:
chore: consistently use sourceFile term

sergiofenoll committed:
feat: link derived file to replacement source file

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3658-source-file-as-main-file

erikap commented:
~http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/63~
http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/64

erikap committed:
Merge branch 'development' into feature/KAS-3658-source-file-as-main-file

erikap committed:
Merge branch 'feature/KAS-3658-source-file-as-main-file' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-3658-source-file-as-main-file

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3658-source-file-as-main-file

sergiofenoll committed:
test: update document widget and viewer tests

sergiofenoll commented:
All tests are green now http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination_2/68/

erikap committed:
Undo Yggdrasil image overwrite for CI
